sdks/python/apache_beam/runners/interactive/pipeline_analyzer.py

```diff
@@ -0,0 +1,372 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""Analyzes and modifies the pipeline that utilize the PCollection cache.
+
+This module is experimental. No backwards-compatibility guarantees.
+"""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import collections
+import copy
+
+import apache_beam as beam
+from apache_beam.portability.api import beam_runner_api_pb2
+from apache_beam.runners.interactive import cache_manager as cache
+
+
+class PipelineAnalyzer(object):
+  def __init__(self, cache_manager, pipeline_proto, underlying_runner,
+               desired_cache_labels=None):
+    self._cache_manager = cache_manager
+    self._pipeline_proto = pipeline_proto
+    self._underlying_runner = underlying_runner
+    self._desired_cache_labels = desired_cache_labels or []
+
+    self._pipeline_proto_to_execute = None
+    self._top_level_referenced_pcollection_ids = None
+    self._top_level_required_transforms = None
+
+    self._analyze_pipeline()
+
+  def _analyze_pipeline(self):
+    pipeline, context = beam.pipeline.Pipeline.from_runner_api(
+        self._pipeline_proto,
+        self._underlying_runner,
+        options=None,
+        return_context=True)
+
+    # context returned from to_runner_api is more informative than that returned
+    # from from_runner_api.
+    _, context = pipeline.to_runner_api(return_context=True)
+
+    pipeline_info = PipelineInfo(self._pipeline_proto.components)
+    self._caches_used = set()
+
+    def _producing_transforms(pcoll_id, leaf=False):
```

Comment:
Does this need to be a nested function?  Can we move this to directly in the class and give it better documentation?